### PR TITLE
Keep line feeds when wrapping text

### PR DESCRIPTION
**Pull request by leomillon**
Hello!

Here is a "fix" of the text wrapping feature that does not reset the `lineWidth` when a `\n` character is found.

I found this using your (really cool) ArgParser documentation helper.
With this help text : `HMAC signature algorithm and base64 encoded secret key.\nAvailable algorithms : [HS256, HS384, HS512]`
I got this kind of result :
```
  --hmac-sign HMAC_ALG SECRET    HMAC signature algorithm and base64 encoded secret key.
                                 Available
                                 algorithms : [HS256, HS384, HS512]
```

But I expected :
```
   --hmac-sign HMAC_ALG SECRET    HMAC signature algorithm and base64 encoded secret key.
                                  Available algorithms : [HS256, HS384, HS512]
```

Thanks a lot for your tools !

### CODE REVIEW

**Comment by xenomachina** [comment]
Thanks for the contribution, and thanks for your patience!

I'm concerned about this breaking up text where people used multi-line strings.

What do you think about making it so that `\n\n` is preserved but `\n` is treated as a space (sort of like markdown does)?

or example, this would turn into two separate paragraphs:

```
        help="""HMAC signature algorithm and base64 encoded secret key.

                Available algorithms : [HS256, HS384, HS512]"""
```

but this would stay as one:

```
        help="""This is one sentence that's just too long for me to fit on one
                line so I had to continue it on the next one."""
```

**Comment by leomillon** [comment]
Yeah good point.

As you suggested to manage line breaks like "markdown paragraph" I tried to treat it the same way : multiple/mix of line breaks/spaces are considered as only one new paragraph.

Example :
```kotlin
"1...\n2...\n\n \n\n\n    \n Hello world!".wrapText(10) shouldBe
"1... 2...\n\nHello\nworld!"
```

Also, as CODEBEAT raised some issues, I change a little bit the function structure (moved most of the body into a dedicated class/methods).

Let me know if it's ok for you :)

File: src/test/kotlin/TextTermTest.kt [hunk 1]
```diff
@@ -99,6 +99,49 @@ class String_WrapTextTest : FunSpec() {
...
+            "1... 2..."
+        }
+
+        test("trailing multiple new-line characters are removed") {
+            "1... 2...\n\n".wrapText(10) shouldBe
+            "1... 2..."
+        }
+
+        test("mix of multiple new-line characters and spaces should be treated as 'markdown paragraphs' (escaped string version)") {
+            "1...\n2...\n\n \n\n\n    \n Hello world!".wrapText(10) shouldBe
```
Comment (xenomachina) [inline src/test/kotlin/TextTermTest.kt]: Could you add another assertion here where none of the newlines are actually adjacent? eg:

    ""1...\n2... \n    \n Hello world!".wrapText(10)

File: src/main/kotlin/TextTerm.kt [hunk 2]
```diff
@@ -53,49 +54,98 @@ fun String.padLinesToWidth(width: Int): String {
...
+        if (Character.isSpaceChar(inputCodePoint) && inputCodePoint != NBSP_CODEPOINT) {
+            // space
+            handleSpace()
+        } else if (inputCodePoint == LINE_FEED_CODEPOINT) {
+            // line feed
+            handleLineFeed()
+        } else {
+            // non-space
+            handleNonSpace(inputCodePoint)
+        }
```
Comment (xenomachina) [inline src/main/kotlin/TextTerm.kt:81]: Thanks for factoring these out!

File: src/main/kotlin/TextTerm.kt [hunk 3]
```diff
@@ -53,49 +54,98 @@ fun String.padLinesToWidth(width: Int): String {
...
+
+internal class TextWrapper(val maxWidth: Int) {
+    private val sb = StringBuilder()
+    private val word = StringBuilder()
+    private var lineWidth = 0
+    private var wordWidth = 0
+    private var newLinesSinceLastWord = 0
+
+    fun addCodePoint(inputCodePoint: Int) {
+
```
Comment (xenomachina) [inline src/main/kotlin/TextTerm.kt]: Please remove this blank line.

**Review by xenomachina — APPROVED** [review]
I had a couple of really minor comments, but overall this looks great.

Thanks for making this improvement!

File: src/main/kotlin/TextTerm.kt (hunk 3 above)
Comment (leomillon) [reply src/main/kotlin/TextTerm.kt]: Done :)

File: src/test/kotlin/TextTermTest.kt (hunk 1 above)
Comment (leomillon) [reply src/test/kotlin/TextTermTest.kt]: Done

**Comment by xenomachina** [comment]
I've merged this in. Thanks again!

**Pull request merged**